In [2]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [3]:
dataset_path="C://sushmitha one drive//Documents//66B8//CNN//PetImages"

In [4]:
#Image Processing
datagen=ImageDataGenerator(
    #Normalize pixels
    rescale=1./255,
    #Data augmentation
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,

    #split dataset
    validation_split=0.2 #validation split used to create train and validation data
)

In [5]:
#Loading the training data
train_data=datagen.flow_from_directory(
    dataset_path,
    target_size=(128,128),
    batch_size=32,
    class_mode='binary',
    subset='training'
 )

Found 20000 images belonging to 2 classes.


In [6]:
# Load validation data 

val_data = datagen.flow_from_directory(
    dataset_path,
    target_size = (128, 128),  # Resize images to 150x150
    batch_size = 32,
    class_mode = 'binary',  # Binary classification (cat vs dog)
    subset = 'validation'  # Use the validation subset
)

Found 4998 images belonging to 2 classes.


In [7]:
print(train_data.class_indices)

{'Cat': 0, 'Dog': 1}


In [8]:
# Building CNN  model
model=keras.Sequential()


In [9]:
#First Convolution Layer


model.add(keras.layers.Conv2D(
    filters=32,
    kernel_size=(3,3),
    activation='relu',
    input_shape=(128,128,3)
))

c:\Users\sushm\anaconda3\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [10]:
#First Pooling layer

model.add(
    keras.layers.MaxPooling2D(
        pool_size=(2,2)
    )
)

In [11]:
# 2 Conv Layer

model.add(keras.layers.Conv2D(
    filters=32,
    kernel_size=(3,3),
    activation='relu'
))

In [12]:
#Second Pooling layer
model.add(
    keras.layers.MaxPooling2D(
        pool_size=(2,2)
    )
)

In [13]:
#Flattern

model.add(keras.layers.Flatten())

In [14]:
#Dense Layer
model.add(
    keras.layers.Dense(
        units=128,
        activation='relu'
    )
)

In [15]:
#output layer
model.add(keras.layers.Dense(
    units=1,
    activation='sigmoid'
))

In [16]:
#View Model architecture

model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 126, 126, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 63, 63, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 61, 61, 32)     │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 30, 30, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 28800)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │     3,686,528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 1)              │           129 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,696,801 (14.10 MB)

 Trainable params: 3,696,801 (14.10 MB)

 Non-trainable params: 0 (0.00 B)

# (None,126,126,channels)

# (Batch size, Height , Width, No of Filters)

Input =128*128*3

filter=3*3

no.of filters=32

output size formula:

output= ((Input-kernel)/(Stride))+1

stride=1

output=((128-3)/1)+1

=126

Height=126

Width=126

chanels--->32

Final output-->(None,126,126,32)

kernel

Formula= (kh*kw*InputChannels+1)*Filters

kernel=3*3

InputChannels:3

Filters:32

Params=(3*3*3+1)*32=896

### Pooling2D

MaxPooling2D((2,2))

Pooling=No learning, No Weights

(only reduces size)

Input=126*126*32

pool=2*2

Formula:126/2=63

output :63*63*32

Final:(None,63,63,32)

## Second Conv2D_1

Input from pooling:(None,63*63*32)

kernel: 3*3

Output: 63-3 +1=61

(None,61,61,32)

param= 9248

### Second Pooling

(None,30,30,32)

param=0

Input=(61,61,32)

Param=0

Pooling=2*2

Floor division:61/2=30

Final:(None,30,30,32)

No Parameters.

##### Flatten

Input : 30 * 30 * 32

Output : 30 * 30 * 32 = 28800 ( Numbers)

Faltten : It will convert 3D ------> 1D Vector (Flatten only reshapes)

##### Dense Layer

Input  = 28800

Neurons = 128

Parameter = (Input * Neurons) + Bias

          = (28800 * 128) + 128

          = 3686400 + 128

          = 3686528

##### Output Layer

Input = 128

Output = 1

Output shape(None,1)

Meaning = Single prediction 

-------> Cat or Dog 

Parameter Calculation :

        = 128 * 1 + 1

        = 129

Input ---> Conv1 ---> Pool1 ---> Conv2 ---> Pool2 ---> Flatten ---> Dense ---> Output









In [17]:

import tensorflow as tf
print(tf.config.list_physical_devices())


[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU')]


In [18]:
# Compile CNN
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

In [19]:
# Train
history=model.fit(
    train_data,
    epochs=10,
    validation_data=val_data
)

Epoch 1/10
549/625 ━━━━━━━━━━━━━━━━━━━━ 57s 762ms/step - accuracy: 0.5794 - loss: 0.7417

c:\Users\sushm\anaconda3\Lib\site-packages\PIL\TiffImagePlugin.py:949: UserWarning: Truncated File Read
  warnings.warn(str(msg))


625/625 ━━━━━━━━━━━━━━━━━━━━ 583s 927ms/step - accuracy: 0.6387 - loss: 0.6351 - val_accuracy: 0.7259 - val_loss: 0.5420
Epoch 2/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 497s 795ms/step - accuracy: 0.7279 - loss: 0.5393 - val_accuracy: 0.7411 - val_loss: 0.5069
Epoch 3/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 467s 740ms/step - accuracy: 0.7606 - loss: 0.4938 - val_accuracy: 0.7579 - val_loss: 0.4936
Epoch 4/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 662s 1s/step - accuracy: 0.7782 - loss: 0.4641 - val_accuracy: 0.7831 - val_loss: 0.4606
Epoch 5/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 427s 682ms/step - accuracy: 0.7928 - loss: 0.4407 - val_accuracy: 0.7915 - val_loss: 0.4369
Epoch 6/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 354s 566ms/step - accuracy: 0.8030 - loss: 0.4233 - val_accuracy: 0.7955 - val_loss: 0.4477
Epoch 7/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 413s 661ms/step - accuracy: 0.8154 - loss: 0.4022 - val_accuracy: 0.8003 - val_loss: 0.4365
Epoch 8/10
625/625 ━━━━━━━━━━━━━━━━━━━━ 397s 636ms/step - accuracy: 0.8218 - loss: 0.3901 

In [20]:
model.save("cnn_pet_classifier.h5")

In [21]:
#Evaluate the model
loss,accuracy=model.evaluate(val_data)
print("Accuracy:",accuracy)


157/157 ━━━━━━━━━━━━━━━━━━━━ 56s 357ms/step - accuracy: 0.8209 - loss: 0.4007
Accuracy: 0.8209283947944641
